### PR TITLE
Grafana-UI: Create fast path in Text component

### LINKED ISSUE #76050: GrafanaUI: Create "fast path" for Text component
We should optimise the Text component to not initialise the `ResizeObserver` and `useEffect` when they're not needed (when truncate prop isn't specified).

e.g.

```tsx
function Text() {
    if (props.truncate) {
        return <TruncatedText {...props} />
    }

    return <SimpleText {...props} />
}

function SimpleText() {
    ...
}

function TruncatedText() {
    const intersectionObserver = useMemo(...)
}
```

### PR DESCRIPTION
**Pull request by krishnemd**
**What is this feature?**

Fixes useEffect and Observer calls when not needed.

**Why do we need this feature?**

Make Text component fast.

**Who is this feature for?**

UI 

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/76050

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/Text/Text.tsx [hunk 1]
```diff
@@ -151,3 +112,83 @@ const getTextStyles = (
...
+  };
+
+  if (isOverflowing && element !== 'span') {
+    return (
+      <Tooltip ref={internalRef} content={getTooltipText(children)}>
+        {createElement(element, {
+          ...restProps,
+          style: undefined,
+          className: styles,
+          ref: internalRef,
```
Comment (L-M-K-B) [inline packages/grafana-ui/src/components/Text/Text.tsx:178]: According to the original code this should be undefined. Is there a reason why you changed it?

**Review by L-M-K-B — COMMENTED** [review]
Hi @chauchausoup, Thank you so much for contributing to Grafana again! 
I had a look at your changes and left a couple of comments directly in the code. 
I ask you to move `getTextStyles` to the end of the file since we try to keep a consistent format for our ui components.

Additionally I would like you to check the component in Storybook locally. I did this and when I set truncate to true it starts flickering. 

https://github.com/grafana/grafana/assets/48948963/d0825c96-fdfa-4471-ae70-b515d81919e3

**Comment by L-M-K-B** [comment]
@joshhunt Since you opened the issue I would like your opinion on whether this PR provides all required changes you envisioned.

File: packages/grafana-ui/src/components/Text/Text.tsx [hunk 2]
```diff
@@ -44,70 +45,30 @@ export const Text = React.forwardRef<HTMLElement, TextProps>(
         [color, textAlignment, truncate, italic, weight, variant, element]
       )
     );
-    const [isOverflowing, setIsOverflowing] = useState(false);
     const internalRef = useRef<HTMLElement>(null);
 
     // wire up the forwarded ref to the internal ref
     useImperativeHandle<HTMLElement | null, HTMLElement | null>(ref, () => internalRef.current);
```
Comment (L-M-K-B) [inline packages/grafana-ui/src/components/Text/Text.tsx:51]: `internalRef` and `useImperativeHandle` is no longer needed here. The ref from `React.forwardRef` can be passed directly to `createElement()` and `TruncatedText`. Instead the `internalRef`/`useImperativeHandle` needs to moved and used in `TruncatedText`.

File: packages/grafana-ui/src/components/Text/Text.tsx [hunk 3]
```diff
@@ -44,70 +45,30 @@ export const Text = React.forwardRef<HTMLElement, TextProps>(
...
-    const getTooltipText = (children: NonNullable<React.ReactNode>) => {
-      if (typeof children === 'string') {
-        return children;
-      }
-      const html = ReactDomServer.renderToStaticMarkup(<>{children}</>);
-      const getRidOfTags = html.replace(/(<([^>]+)>)/gi, '');
-      return getRidOfTags;
-    };
-    // A 'span' is an inline element therefore it can't be truncated
-    // and it should be wrapped in a parent element that is the one that will show the tooltip
```
Comment (L-M-K-B) [inline packages/grafana-ui/src/components/Text/Text.tsx:101]: This comment should be preserved in the TruncatedText component.

**Review by L-M-K-B — CHANGES_REQUESTED** [review]

**Comment by chauchausoup** [comment]
Hello @L-M-K-B 
Thanks for your review.
I have tried to cover most of your comment on this recent commits. 
Here I have decoupled a separate `TruncatedText` component in terms of component tree and in terms of parameters its depending with the parent `Text` component. 
Let me know if any changes needs to be made on the PR.

**Review by L-M-K-B — COMMENTED** [review]
You are nearly there. There are only two small changes left.
Additionally I need to ask you to get the latest version of main from the Grafana repo and rebase your feature branch. We did some changes and the tests in our build pipeline will fail otherwise.

**Comment by chauchausoup** [comment]
@L-M-K-B just made the changes. Let me know.

**Review by L-M-K-B — APPROVED** [review]

**Comment by L-M-K-B** [comment]
Hi @chauchausoup, Thanks a lot for your support. Unfortunately e2e tests in our [CI pipeline](https://drone.grafana.net/grafana/grafana/144244/1/1) are still failing. That's why I need to ask you to merge the current version of main again.

**Pull request merged**